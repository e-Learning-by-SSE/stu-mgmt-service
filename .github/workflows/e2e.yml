name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
