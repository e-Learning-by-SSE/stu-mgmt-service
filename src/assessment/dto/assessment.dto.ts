import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
