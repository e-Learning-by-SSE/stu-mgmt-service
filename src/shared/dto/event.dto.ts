export class EventDto {
	event: string;
	timestamp: Date;
	payload?: Record<string, unknown>;
}
