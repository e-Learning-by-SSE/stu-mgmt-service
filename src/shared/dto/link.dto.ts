export class LinkDto {
	url: string;
	name: string;
}
