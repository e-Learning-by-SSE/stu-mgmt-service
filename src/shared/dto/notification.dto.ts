import { AssignmentId } from "../../course/entities/assignment.entity";
import { CourseId } from "../../course/entities/course.entity";
import { GroupId } from "../../course/entities/group.entity";
import { Event } from "../../course/events/events";
import { UserId } from "../entities/user.entity";

export class NotificationDto {
	event: Event;
	courseId: CourseId;
	assignmentId?: AssignmentId;
	groupId?: GroupId;
	userId?: UserId;
	payload?: Record<string, unknown>;
}
