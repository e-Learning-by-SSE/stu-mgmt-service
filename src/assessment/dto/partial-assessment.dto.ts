import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
