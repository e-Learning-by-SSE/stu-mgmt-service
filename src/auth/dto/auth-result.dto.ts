import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
