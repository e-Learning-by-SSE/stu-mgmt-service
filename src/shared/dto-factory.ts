import { AssessmentDto } from "../assessment/dto/assessment.dto";
import { Assessment } from "../assessment/entities/assessment.entity";
import { AssignmentDto } from "../course/dto/assignment/assignment.dto";
import { AdmissionCriteriaDto } from "../course/dto/course-config/admission-criteria.dto";
import { CourseConfigDto } from "../course/dto/course-config/course-config.dto";
import { GroupSettingsDto } from "../course/dto/course-config/group-settings.dto";
import { ParticipantDto } from "../course/dto/course-participant/participant.dto";
import { CourseDto } from "../course/dto/course/course.dto";
import { GroupDto } from "../course/dto/group/group.dto";
import { AdmissionCriteria } from "../course/entities/admission-criteria.entity";
import { Assignment } from "../course/entities/assignment.entity";
import { CourseConfig } from "../course/entities/course-config.entity";
import { Course } from "../course/entities/course.entity";
import { GroupSettings } from "../course/entities/group-settings.entity";
import { Group } from "../course/entities/group.entity";
import { UserDto } from "./dto/user.dto";
import { User } from "./entities/user.entity";
import { CourseRole } from "./enums";
import { toDtos } from "./interfaces/to-dto.interface";

export abstract class DtoFactory {
	static createCourseDto(course: Course): CourseDto {
		return {
			id: course.id,
			shortname: course.shortname,
			semester: course.semester,
			title: course.title,
			isClosed: course.isClosed,
			links: course.links ?? undefined,
			groupSettings: course.config?.groupSettings?.toDto(),
			admissionCriteria: course.config?.admissionCriteria?.toDto()
		};
	}

	static createCourseConfigDto(config: CourseConfig, includePrivileged = false): CourseConfigDto {
		const configDto: CourseConfigDto = {};
		if (config.admissionCriteria)
			configDto.admissionCriteria = this.createAdmissionCriteriaDto(config.admissionCriteria);

		if (config.groupSettings)
			configDto.groupSettings = this.createGroupSettingsDto(config.groupSettings);

		if (includePrivileged) {
			configDto.password = config.password;
		}
		return configDto;
	}

	static createGroupSettingsDto(settings: GroupSettings): GroupSettingsDto {
		return {
			allowGroups: settings.allowGroups,
			nameSchema: settings.nameSchema,
			sizeMin: settings.sizeMin,
			sizeMax: settings.sizeMax,
			selfmanaged: settings.selfmanaged,
			autoJoinGroupOnCourseJoined: settings.autoJoinGroupOnCourseJoined,
			mergeGroupsOnAssignmentStarted: settings.mergeGroupsOnAssignmentStarted
		};
	}

	static createAdmissionCriteriaDto(criteria: AdmissionCriteria): AdmissionCriteriaDto {
		return criteria.admissionCriteria;
	}

	static createUserDto(user: User, options?: { removeEmail: boolean }): UserDto {
		const userDto: UserDto = {
			id: user.id,
			matrNr: user.matrNr,
			email: options?.removeEmail ? undefined : user.email,
			username: user.username,
			displayName: user.displayName,
			role: user.role
		};

		// Add relational data, if available
		if (user.participations) {
			if (user.participations.length == 0) {
				userDto.courses = [];
			} else if (user.participations[0].course) {
				userDto.courses = user.participations.map(rel => this.createCourseDto(rel.course));
			}
		}

		return userDto;
	}

	static createGroupDto(group: Group, options?: { includePassword: boolean }): GroupDto {
		const groupDto: GroupDto = {
			id: group.id,
			name: group.name,
			isClosed: group.isClosed,
			password: options?.includePassword ? group.password : undefined,
			hasPassword: !!group.password
		};

		if (group.userGroupRelations) {
			groupDto.size = group.userGroupRelations.length;
			groupDto.members = [];

			if (group.userGroupRelations.length && group.userGroupRelations[0].participant) {
				groupDto.members = toDtos(group.userGroupRelations.map(x => x.participant));
			}
		}

		if (group.history) {
			groupDto.history = group.history.map(event => event.toDto());
		}

		return groupDto;
	}

	static createAssignmentDto(assignment: Assignment, isLecturer = false): AssignmentDto {
		const assignmentDto: AssignmentDto = {
			id: assignment.id,
			name: assignment.name,
			state: assignment.state,
			startDate: assignment.startDate ?? undefined,
			endDate: assignment.endDate ?? undefined,
			comment: assignment.comment ?? undefined,
			links: assignment.links ?? undefined,
			type: assignment.type,
			points: assignment.points,
			bonusPoints: assignment.bonusPoints ?? undefined,
			collaboration: assignment.collaboration,
			configs: assignment.configs ?? undefined
		};

		// Show tool configurations only to lecturers and hide them for students
		if (!isLecturer) {
			assignmentDto.configs = undefined;
		}

		return assignmentDto;
	}

	static createAssessmentDto(assessment: Assessment): AssessmentDto {
		const assessmentDto: AssessmentDto = {
			id: assessment.id,
			assignmentId: assessment.assignmentId,
			groupId: assessment.groupId ?? undefined,
			isDraft: assessment.isDraft,
			achievedPoints: assessment.achievedPoints,
			comment: assessment.comment ?? undefined,
			creatorId: assessment.creatorId,
			lastUpdatedById: assessment.lastUpdatedById ?? undefined,
			creationDate: assessment.creationDate,
			updateDate: assessment.updateDate,
			partialAssessments: assessment.partialAssessments?.map(p => p.toDto())
		};

		if (!assessmentDto.isDraft && assessmentDto.partialAssessments?.length > 0) {
			assessmentDto.partialAssessments = assessmentDto.partialAssessments.filter(
				p => !p.draftOnly
			);
		}

		if (assessment.assignment) {
			assessmentDto.assignment = this.createAssignmentDto(assessment.assignment);
		}

		// If assessment belongs to a single student
		if (assessment.assessmentUserRelations?.length == 1) {
			assessmentDto.userId = assessment.assessmentUserRelations[0].userId;

			// If full user was included
			if (assessment.assessmentUserRelations[0].user) {
				const user = assessment.assessmentUserRelations[0].user;
				assessmentDto.participant = {
					userId: user.id,
					matrNr: user.matrNr,
					username: user.username,
					displayName: user.displayName,
					email: user.email,
					role: CourseRole.STUDENT
				};
			}
		}

		// If creator was loaded
		if (assessment.creator)
			assessmentDto.creator = this.createUserDto(assessment.creator, { removeEmail: true });
		if (assessment.lastUpdatedBy)
			assessmentDto.lastUpdatedBy = this.createUserDto(assessment.lastUpdatedBy, {
				removeEmail: true
			});

		if (assessment.group) {
			assessmentDto.group = this.createGroupDto(assessment.group);
			assessmentDto.group.members = assessment.assessmentUserRelations?.map(rel => {
				const participant: ParticipantDto = {
					// TODO: GroupId missing
					role: CourseRole.STUDENT,
					userId: rel.userId,
					matrNr: rel.user.matrNr,
					username: rel.user.username,
					displayName: rel.user.displayName,
					email: rel.user.email
				};
				return participant;
			});
			assessmentDto.group.size = assessmentDto.group.members?.length;
		}

		return assessmentDto;
	}
}
