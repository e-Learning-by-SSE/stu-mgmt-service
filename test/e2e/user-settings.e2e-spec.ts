import { Language } from "../../src/shared/language";
import { UserSettingsDto } from "../../src/user/dto/user-settings.dto";
import { USER_SETTINGS_MOCK } from "../mocks/user-settings.mock";
import { TestSetup } from "../utils/e2e";
import { copy } from "../utils/object-helper";

const route = (userId: string) => `/users/${userId}/settings`;

describe("UserSettings E2E", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	describe("POST-REQUESTS of UserSettingsController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(PUT) /users/{userId}/settings", () => {
			const userId = USER_SETTINGS_MOCK[0].userId;
			const settings = copy(USER_SETTINGS_MOCK[0].userSettings);

			settings.allowEmails = !settings.allowEmails;
			settings.language = Language.EN;
			settings.blacklistedEvents = null;

			it("Updates user settings", () => {
				return setup
					.request()
					.put(route(userId))
					.send(settings)
					.expect(200)
					.expect(({ body }) => {
						const result = body as UserSettingsDto;
						expect(result.allowEmails).toEqual(settings.allowEmails);
						expect(result.language).toEqual(settings.language);
						expect(result.blacklistedEvents).toEqual(settings.blacklistedEvents);
					});
			});
		});
	});

	describe("GET-REQUESTS of UserSettingsController (e2e)", () => {
		beforeAll(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(GET) /users/{userId}/settings", () => {
			it("Returns user settings", () => {
				const userId = USER_SETTINGS_MOCK[1].userId;

				return setup
					.request()
					.get(route(userId))
					.expect(200)
					.expect(({ body }) => {
						expect(body).toMatchSnapshot();
					});
			});
		});
	});
});
