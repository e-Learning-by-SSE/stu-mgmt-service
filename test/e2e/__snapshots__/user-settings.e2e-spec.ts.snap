// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings E2E GET-REQUESTS of UserSettingsController (e2e) (GET) /users/{userId}/settings Returns user settings 1`] = `
Object {
  "allowEmails": true,
  "blacklistedEvents": Object {
    "ASSIGNMENT_STATE_CHANGED": true,
  },
  "language": "EN",
}
`;
